N°3863 - exec.php : allow subdirectories in the page parameter

piRGoif opened the pull request:
For now, exec.php can only call pages at the root level of modules. This is caused by basename() calls on both page and module parameters.

This PR brings the possibility to call pages in the whole module hierarchy, still using `\utils::RealPath` to ensure we're not trying to access something outside of the module.

piRGoif committed:
Fix exec.php cannot load page with subdir
For example `/pages/exec.php?exec_module=itop-hub-connector&exec_page=test%2Ftest-landing.html&exec_env=production&operation=install`

piRGoif committed:
Multiple dirs only used if symlink detected

Hipska commented on pages/exec.php:
```diff
@@ -25,32 +25,64 @@
+if (is_link($sPageEnvFullPath)) {
+	// in case module was compiled to symlink, we need to check against real linked path as symlink is resolved
+	$aPossibleBasePaths = [
+		APPROOT.'datamodels/2.x', // warning, won't work for datamodels/1.x for example !
```
Not `3.x` for new major iTop release?

piRGoif replied:
No we didn't changed the directory name, see https://github.com/Combodo/iTop/tree/develop/datamodels
Indeed there was just some little refactoring (like the itop-structure module creation) but no revolution : same iTop objects and attributes.

Molkobain replied:
This means that the extensions made using this feature won't work on the iTop used internally at Combodo (poke @v-dumas ) and on other legacy packages. It might worth it to test for both 1.x and 2.x folders?

v-dumas replied:
Do not put any code inside iTop which would not be compatible with legacy packages, that would be too much of a nightmare for the legacies...

Molkobain replied:
```suggestion
APPROOT.'datamodels/1.x', // Compatibility with legacy packages
APPROOT.'datamodels/2.x',
```

piRGoif replied:
Good point, thanks ! This should be solved by using \ModelFactory::GetRootDirs

Hipska replied:
But it still isn't

piRGoif replied:
Should be solved in 53a2fe2e

piRGoif commented:
Dev team review conclusions : 
* [ ] the new behavior should be explicitly enabled (white list of the modules, or better an option in the module.*.php file !)
* [ ] there is already an existing provider for the extension locations : \ModelFactory::GetRootDirs
* [ ] does a test exist for utils::RealPath ?
* [ ] add a test for CheckPageExists !

Molkobain commented:
Did you ask @dflaven about why he didn't enable this in the beginning? It might have been on purpose.

piRGoif commented:
I talk about this with @rquetiez who made this originally.
This code is very old, and now we have \utils::RealPath \o/

Molkobain commented on pages/exec.php:
```diff
@@ -25,32 +25,64 @@
 	echo "Missing argument 'exec_module'";
 	exit;
 }
-$sModule = basename($sModule); // protect against ../.. ...
```
Did we loose this protection?
I understand the need to drill through the folders of the module, but we *must not* be able to execute something outside of the iTop folder (eg. `../../../home/foo/malicious.php`)

piRGoif replied:
We are now calling CheckPageExists function, which calls utils::RealPath
This later ensure we won't allow to access files in the directories above in the hierarchy

Molkobain replied:
Thanks! 👌
I saw the call of `utils::RealPath` but didn't noticed that it was actually checking that the filepath was contained in the basepath; preventing that kind a vulnerability; my bad 🙈

piRGoif replied:
No prb, better check twice this kind of things O:)

Molkobain commented on pages/exec.php:
```diff
@@ -25,32 +25,64 @@
 	echo "Missing argument 'exec_page'";
 	exit;
 }
-$sPage = basename($sPage); // protect against ../.. ...
```
Same comment as above

piRGoif commented:
Rebased branch

piRGoif committed:
Compatibility with symlinks + legacy builds
We were browsing for source dir in datamodels/2.x but not datamodels/1.x... This is solved by using the 'source_dir' config parameter (generated by the compiler)

Molkobain commented:
Accepted in technical team review

Molkobain commented:
Accepted in functional team review.

piRGoif commented:
This new feature needs to be documented, but this will be done later on, and the follow-up will be done using the Combodo private ticketing system.
I'm merging !
